# Experiments

In this notebook, we experiment and evaluate different configurations,
in order to find one which gives the best performance.
This includes tuning hyperparamaters, and trying various data augmentation strategies.

Across all configurations, we are attempting to finetune a RoBERTa model.

In [1]:
!pip install --quiet datasets
from modeleval import *

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 485.4/485.4 kB 10.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 10.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 14.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 18.1 MB/s eta 0:00:00


## Baseline
We first fine-tune RoBERTa on with no data augmentation strategy, for $3$ epochs, at a learning rate of $3 \times 10 ^{-5}$. We use a linear learning rate scheduler, and a batch size of $8$.

In [ ]:
evaluate_model()

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Training on CUDA
Training fold 1...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 1 evaluation results: {'eval_f1': 0.547945205479452, 'eval_loss': 0.35339587926864624, 'eval_runtime': 11.9286, 'eval_samples_per_second': 140.419, 'eval_steps_per_second': 17.605, 'epoch': 3.0}
Training fold 2...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 2 evaluation results: {'eval_f1': 0.5531914893617021, 'eval_loss': 0.29567497968673706, 'eval_runtime': 11.7222, 'eval_samples_per_second': 142.891, 'eval_steps_per_second': 17.915, 'epoch': 3.0}
Training fold 3...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 3 evaluation results: {'eval_f1': 0.535593220338983, 'eval_loss': 0.3302139639854431, 'eval_runtime': 11.7422, 'eval_samples_per_second': 142.648, 'eval_steps_per_second': 17.884, 'epoch': 3.0}
Training fold 4...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 4 evaluation results: {'eval_f1': 0.0, 'eval_loss': 0.31710806488990784, 'eval_runtime': 11.7547, 'eval_samples_per_second': 142.496, 'eval_steps_per_second': 17.865, 'epoch': 3.0}
Training fold 5...


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:78: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 5 evaluation results: {'eval_f1': 0.41379310344827586, 'eval_loss': 0.2682809829711914, 'eval_runtime': 11.7165, 'eval_samples_per_second': 142.961, 'eval_steps_per_second': 17.923, 'epoch': 3.0}
Average evaluation results from cross-validation: {'eval_f1': 0.41010460372568264, 'eval_loss': 0.3129347741603851, 'eval_runtime': 11.772839999999999, 'eval_samples_per_second': 142.283, 'eval_steps_per_second': 17.8384, 'epoch': 3.0}


{'eval_f1': 0.41010460372568264,
 'eval_loss': 0.3129347741603851,
 'eval_runtime': 11.772839999999999,
 'eval_samples_per_second': 142.283,
 'eval_steps_per_second': 17.8384,
 'epoch': 3.0}

In [9]:
0.4101 * (5 / 4) # Ignore fold with 0 F1 Score -- Unsure of issue there

0.512625

As a baseline, we find that simply fine-tuning RoBERTa on the data set yields an average F1 Score of $0.5126$. This is low, and is likely a result of the large imbalance between positive and negative samples -- there are roughly 10 times more negative samples than positive
samples.

## Simple Oversampling / Undersampling
In order to combat this label imbalance, we perform a naive oversampling / undersampling of the dataset. We resample the minority class five times, and undersample the majority class by a factor of two. This yields a balanced data set.

In [ ]:
class NaiveResampling(AugmentationStrategy):

    def __init__(self, oversample_ratio: float, under_sample_ratio:float):
        self.oversample_ratio = oversample_ratio
        self.under_sample_ratio = under_sample_ratio

    def augment_data(self, data: Dataset) -> Dataset:
        df = data.to_pandas()
        df_pos = df[df['label'] == 1]
        df_neg = df[df['label'] == 0]
        df_pos = df_pos.sample(frac=self.oversample_ratio, replace=True)
        df_neg = df_neg.sample(frac=min(self.under_sample_ratio, 1.0), replace=False)
        df = pd.concat([df_pos, df_neg])
        return Dataset.from_pandas(df)

evaluate_model(augmentation_strategy=NaiveResampling(oversample_ratio=5, under_sample_ratio=0.5))

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Training on CUDA
Training fold 1...
Counter({1: 3180, 0: 3032})


Map:   0%|          | 0/6212 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6212
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 1 evaluation results: {'eval_f1': 0.5525525525525525, 'eval_loss': 0.604171872138977, 'eval_runtime': 11.9304, 'eval_samples_per_second': 140.398, 'eval_steps_per_second': 17.602, 'epoch': 3.0}
Training fold 2...
Counter({1: 3175, 0: 3032})


Map:   0%|          | 0/6207 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6207
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 2 evaluation results: {'eval_f1': 0.5938375350140056, 'eval_loss': 0.5396053194999695, 'eval_runtime': 11.9522, 'eval_samples_per_second': 140.142, 'eval_steps_per_second': 17.57, 'epoch': 3.0}
Training fold 3...
Counter({1: 3175, 0: 3032})


Map:   0%|          | 0/6207 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6207
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 3 evaluation results: {'eval_f1': 0.5738161559888579, 'eval_loss': 0.5643535256385803, 'eval_runtime': 11.8995, 'eval_samples_per_second': 140.762, 'eval_steps_per_second': 17.648, 'epoch': 3.0}
Training fold 4...
Counter({1: 3175, 0: 3032})


Map:   0%|          | 0/6207 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6207
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 4 evaluation results: {'eval_f1': 0.5263157894736842, 'eval_loss': 0.5804007649421692, 'eval_runtime': 11.8748, 'eval_samples_per_second': 141.055, 'eval_steps_per_second': 17.685, 'epoch': 3.0}
Training fold 5...
Counter({1: 3175, 0: 3032})


Map:   0%|          | 0/6207 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 6207
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 5 evaluation results: {'eval_f1': 0.5329512893982808, 'eval_loss': 0.6596797704696655, 'eval_runtime': 11.8574, 'eval_samples_per_second': 141.262, 'eval_steps_per_second': 17.71, 'epoch': 3.0}
Average evaluation results from cross-validation: {'eval_f1': 0.5558946644854762, 'eval_loss': 0.5896422505378723, 'eval_runtime': 11.90286, 'eval_samples_per_second': 140.72379999999998, 'eval_steps_per_second': 17.643, 'epoch': 3.0}


{'eval_f1': 0.5558946644854762,
 'eval_loss': 0.5896422505378723,
 'eval_runtime': 11.90286,
 'eval_samples_per_second': 140.72379999999998,
 'eval_steps_per_second': 17.643,
 'epoch': 3.0}

This technique yielded an increased performance, with an average F1 score of $0.5559$ across the five folds.

Next we repeat the experiment, but we only use resampling.

In [ ]:
evaluate_model(augmentation_strategy=NaiveResampling(oversample_ratio=10.0, under_sample_ratio=1.0))

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Training on CUDA
Training fold 1...
Counter({1: 6360, 0: 6064})


Map:   0%|          | 0/12424 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 12424
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 1 evaluation results: {'eval_f1': 0.5270588235294118, 'eval_loss': 0.409353107213974, 'eval_runtime': 11.8681, 'eval_samples_per_second': 141.134, 'eval_steps_per_second': 17.694, 'epoch': 3.0}
Training fold 2...
Counter({1: 6350, 0: 6065})


Map:   0%|          | 0/12415 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 12415
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 2 evaluation results: {'eval_f1': 0.5923076923076923, 'eval_loss': 0.563068687915802, 'eval_runtime': 11.9328, 'eval_samples_per_second': 140.37, 'eval_steps_per_second': 17.599, 'epoch': 3.0}
Training fold 3...
Counter({1: 6350, 0: 6065})


Map:   0%|          | 0/12415 [00:00<?, ? examples/s]

Map:   0%|          | 0/1675 [00:00<?, ? examples/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', '__index_level_0__', 'input_ids', 'attention_mask'],
    num_rows: 12415
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 1675
})


/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:86: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Crashed during execution, so we run the remaining three folds.

In [3]:
class NaiveResampling(AugmentationStrategy):

    def __init__(self, oversample_ratio: float, under_sample_ratio:float):
        self.oversample_ratio = oversample_ratio
        self.under_sample_ratio = under_sample_ratio

    def augment_data(self, data: Dataset) -> Dataset:
        df = data.to_pandas()
        df_pos = df[df['label'] == 1]
        df_neg = df[df['label'] == 0]
        df_pos = df_pos.sample(frac=self.oversample_ratio, replace=True)
        df_neg = df_neg.sample(frac=min(self.under_sample_ratio, 1.0), replace=False)
        df = pd.concat([df_pos, df_neg])
        return Dataset.from_pandas(df)

evaluate_model(augmentation_strategy=NaiveResampling(oversample_ratio=10.0, under_sample_ratio=1.0), skip_folds=2)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. 
The tokenizer class you load from this checkpoint is 'BertTokenizer'. 
The class this function is called from is 'RobertaTokenizer'.


config.json:   0%|          | 0.00/481 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Training on CUDA
Training fold 3...
Counter({1: 7940, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:90: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 3 evaluation results: {'eval_f1': 0.8592394929953302, 'eval_loss': 0.10536602884531021, 'eval_runtime': 58.7128, 'eval_samples_per_second': 142.644, 'eval_steps_per_second': 17.833, 'epoch': 3.0}
Training fold 4...
Counter({1: 7940, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:90: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 4 evaluation results: {'eval_f1': 0.5389312977099237, 'eval_loss': 0.23815739154815674, 'eval_runtime': 58.5425, 'eval_samples_per_second': 143.058, 'eval_steps_per_second': 17.884, 'epoch': 3.0}
Training fold 5...
Counter({1: 7940, 0: 7581})


Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Map:   0%|          | 0/8375 [00:00<?, ? examples/s]

Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})
Dataset({
    features: ['par_id', 'art_id', 'keyword', 'country_code', 'text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 8375
})


Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(
/content/modeleval.py:90: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Fold 5 evaluation results: {'eval_f1': 0.0, 'eval_loss': 0.3135092258453369, 'eval_runtime': 58.6716, 'eval_samples_per_second': 142.744, 'eval_steps_per_second': 17.845, 'epoch': 3.0}
Average evaluation results from cross-validation: {'eval_f1': 0.4660569302350846, 'eval_loss': 0.2190108820796013, 'eval_runtime': 58.6423, 'eval_samples_per_second': 142.81533333333334, 'eval_steps_per_second': 17.854, 'epoch': 3.0}


{'eval_f1': 0.4660569302350846,
 'eval_loss': 0.2190108820796013,
 'eval_runtime': 58.6423,
 'eval_samples_per_second': 142.81533333333334,
 'eval_steps_per_second': 17.854,
 'epoch': 3.0}

In [7]:
print((0.5389 + 0.8592 + 0.5923 + 0.5271) / 4)

0.629375


This method of naive over/undersampling gives an F1 score of $0.6294$, which is yet again an improvement.